_P112_Power.ino

```diff
@@ -120,115 +182,128 @@ boolean Plugin_112(byte function, struct EventStruct *event, String& string)
 \*********************************************************************************************/
 void Plugin_112_idleusage(byte Index)
 {
-  unsigned long PulseTime=millis() - Plugin_112_pulseTimePrevious[Index];
-  if(PulseTime > (Settings.TaskDeviceTimer[Index] * 1000) &&        //More than $device_delay passed since last pulse
-     PulseTime > Plugin_112_pulseTime[Index] ) {                    //More than last pulse interval
-
-    // Let's prevent divison by zero
-    if(Settings.TaskDevicePluginConfig[Index][1]==0) {
-      Settings.TaskDevicePluginConfig[Index][1]=1000; // if not configged correctly prevent crashes and set it to 1000 as default value.
-    }
-    // WH = =3600000/[pulses per kwh]/[time since last pulse (ms)]
-    Plugin_112_pulseUsage[Index] = (3600000000./Settings.TaskDevicePluginConfig[Index][1])/PulseTime;
+  unsigned long PulseTime = millis() - Plugin_112_pulseTimePrevious[Index];
+  if(PulseTime > (Settings.TaskDeviceTimer[Index] * 1000) &&  // More than $device_delay passed since last pulse
+     Plugin_112_pulseTimePrevious[Index] > 0 &&               // Must have at least one pulse received
+     PulseTime > Plugin_112_pulseTime[Index] ) {              // More than last pulse interval
+    // WH = 3600000 / [pulses per kwh] / [time since last pulse (ms)]
+    Plugin_112_pulseUsage[Index] = (3600000000. / Settings.TaskDevicePluginConfig[Index][IndexPulsesPerkWh]) / PulseTime;
   }
 }
 
 
 /*********************************************************************************************\
- * Check Pulse Counters (called from irq handler)
+* Check Pulse Counters (called from irq handler)
 \*********************************************************************************************/
 void Plugin_112_pulsecheck(byte Index)
 {
-  unsigned long PulseTime=millis() - Plugin_112_pulseTimePrevious[Index];
-  if(PulseTime > Settings.TaskDevicePluginConfig[Index][0]) // check with debounce time for this task
-    {
-      Plugin_112_pulseTimePrevious[Index]=millis(); // moved this up as the operations below might take some time
-      Plugin_112_pulseCounter[Index]++;
-      Plugin_112_pulseTotalCounter[Index]++;
-      Plugin_112_pulseTime[Index] = PulseTime;
+  noInterrupts(); // s0170071: avoid nested interrups due to bouncing.
+  // s0170071: the following gives a glitch if millis() rolls over (every 50 days) and there is a bouncing to be avoided at the exact same
+  // time. Very rare.
+  // Alternatively there is timePassedSince(Plugin_112_bounceTimePrevious[Index]); but this is not in IRAM at this time, so do not use in a
+  // ISR!
```

Comment:
Hmm I see no reason why `timePassedSince` cannot have the `ICACHE_RAM_ATTR` attribute.

Reply:
I did not look into this, these comments are straight from P003 and I left them in for completeness. It will not make much difference for this plugin I think. Do you think it is worthwhile to explore?

Reply:
Well if it is part of P003, then we will come across it for that plugin.